metatomic-torch/src/io/io.cpp

```diff
@@ -0,0 +1,244 @@
+  ensure(zr.has("positions.npy") && zr.has("cell.npy") &&
+         zr.has("types.npy") && zr.has("pbc.npy"),
+         "File does not contain a valid System object (.npy core missing)");
+
+  // Load core arrays
+  auto pos   = io::npy_read(zr.read("positions.npy")).contiguous().to(torch::kFloat64).cpu();
+  auto cell  = io::npy_read(zr.read("cell.npy")).contiguous().to(torch::kFloat64).cpu();
+  auto types = io::npy_read(zr.read("types.npy")).contiguous().cpu();
+  auto pbc   = io::npy_read(zr.read("pbc.npy")).contiguous().to(torch::kBool).cpu();
+
+  // Construct System (constructor: types, positions, cell, pbc)
+  System system = torch::make_intrusive<SystemHolder>(types, pos, cell, pbc);
+
+  // Neighbor lists: find all "pairs/<idx>/options.json"
+  std::vector<std::string> names = zr.list();
+  struct NLRec { size_t idx; std::string options_path; std::string data_path; };
+  std::vector<NLRec> recs;
+
+  for (const auto& name : names) {
+    if (starts_with(name, "pairs/") && ends_with(name, "/options.json")) {
+      const auto after_pairs = name.substr(6); // "<idx>/options.json"
+      auto slash = after_pairs.find('/');
+      ensure(slash != std::string::npos, "malformed neighbor list path");
+      auto idx_str = after_pairs.substr(0, slash);
+      size_t idx = static_cast<size_t>(std::stoul(idx_str));
+      recs.push_back({ idx, name, "pairs/" + std::to_string(idx) + "/data.mts" });
+    }
+  }
+  std::sort(recs.begin(), recs.end(), [](const NLRec& a, const NLRec& b){ return a.idx < b.idx; });
+
+  for (const auto& r : recs) {
+    // options.json -> NeighborListOptions
+    auto options_bytes = zr.read(r.options_path);
+    std::string json(reinterpret_cast<const char*>(options_bytes.data()), options_bytes.size());
+    NeighborListOptions opts = NeighborListOptionsHolder::from_json(json);
+
+    // data.mts -> TensorBlock
+    auto data_bytes = zr.read(r.data_path);
+    auto bytes_sp = std::make_shared<std::vector<uint8_t>>(data_bytes);
+    auto t = torch::from_blob(
+        bytes_sp->data(),
+        { static_cast<long>(bytes_sp->size()) },
+        // deleter: called when tensor storage is released
+        [bytes_sp](void* /*unused*/) mutable {
+            // bytes_sp will be destroyed here; vector frees its memory.
+            bytes_sp.reset();
+        },
+        torch::TensorOptions().dtype(torch::kUInt8).device(torch::kCPU)
+    );
+
+    TensorBlock block = metatensor_torch::load_block_buffer(t);
+
+    system->add_neighbor_list(opts, block);
+  }
+
+  // Extra data under data/*.mts
+  for (const auto& name : names) {
+    if (starts_with(name, "data/") && ends_with(name, ".mts")) {
+      const auto stem = name.substr(5); // "<key>.mts"
+      const auto dot = stem.rfind('.');
+      ensure(dot != std::string::npos, "malformed extra data path");
+      const auto key = stem.substr(0, dot);
+
+      auto bytes = zr.read(name);
+      auto bytes_sp = std::make_shared<std::vector<uint8_t>>(bytes);
```

Comment:
```suggestion
      auto bytes_sp = std::make_shared<std::vector<uint8_t>>(std::move(bytes));
```

Otherwise there is one extra copy. `clangd` can help to show you this kind of performance tricks in vscode, I can help you set it up if you want.